node/insecure_validator_manager.go

```diff
@@ -4,38 +4,53 @@
 package node
 
 import (
+	"go.uber.org/zap"
+
 	"github.com/ava-labs/avalanchego/ids"
 	"github.com/ava-labs/avalanchego/snow/networking/router"
 	"github.com/ava-labs/avalanchego/snow/validators"
-	"github.com/ava-labs/avalanchego/utils/constants"
+	"github.com/ava-labs/avalanchego/utils/logging"
 	"github.com/ava-labs/avalanchego/version"
 )
 
 type insecureValidatorManager struct {
 	router.Router
-	vdrs   validators.Set
-	weight uint64
+	log      logging.Logger
+	subnetID ids.ID
+	vdrs     validators.Manager
+	weight   uint64
 }
 
 func (i *insecureValidatorManager) Connected(vdrID ids.NodeID, nodeVersion *version.Application, subnetID ids.ID) {
-	if constants.PrimaryNetworkID == subnetID {
+	if i.subnetID == subnetID {
 		// Sybil protection is disabled so we don't have a txID that added the
 		// peer as a validator. Because each validator needs a txID associated
 		// with it, we hack one together by padding the nodeID with zeroes.
 		dummyTxID := ids.Empty
 		copy(dummyTxID[:], vdrID[:])
 
-		// Add will only error here if the total weight of the set would go over
-		// [math.MaxUint64]. In this case, we will just not mark this new peer
-		// as a validator.
-		_ = i.vdrs.Add(vdrID, nil, dummyTxID, i.weight)
+		err := i.vdrs.AddStaker(i.subnetID, vdrID, nil, dummyTxID, i.weight)
+		if err != nil {
+			i.log.Error("failed to add validator",
+				zap.Stringer("nodeID", vdrID),
+				zap.Stringer("subnetID", i.subnetID),
+				zap.Error(err),
+			)
+		}
 	}
 	i.Router.Connected(vdrID, nodeVersion, subnetID)
 }
 
 func (i *insecureValidatorManager) Disconnected(vdrID ids.NodeID) {
 	// RemoveWeight will only error here if there was an error reported during
 	// Add.
-	_ = i.vdrs.RemoveWeight(vdrID, i.weight)
+	err := i.vdrs.RemoveWeight(i.subnetID, vdrID, i.weight)
+	if err != nil {
+		i.log.Error("failed to add validator",
+			zap.Stringer("nodeID", vdrID),
+			zap.Stringer("subnetID", i.subnetID),
+			zap.Error(err),
+		)
+	}
```

Comment:
This isn't correct unless `i.subnetID` is the primary network id - because `Connected` may have been called with a different subnetID (say the primary network ID) but never called with `i.subnetID` so when `Disconnected` is called the weight will be removed incorrectly

Reply:
removed it, thanks.